resolver/models/substance.py

```diff
@@ -28,6 +29,41 @@ class Substance(db.Model):
     casrn = index_property("identifiers", "casrn", default=None)
     preferred_name = index_property("identifiers", "preferred_name", default=None)
 
+    @hybrid_method
+    def get_matches(self, searchterm=None):
+        if bool(searchterm):
+            # search the identifiers for the term and score them
+            matchlist = {}
+            for id_name in ["casrn", "preferred_name", "display_name", "compound_id"]:
+                # an exact match against a top-level identifier
+                # yields a 1.0 score
+                if self.identifiers[id_name] == searchterm:
+                    matchlist[id_name] = 1
+            if self.identifiers["synonyms"]:
+                synonyms = self.identifiers["synonyms"]
+                # a match against a synonym identifier
+                # yields whatever the weight of the synonym was
+                # set to
+                for synonym in synonyms:
+                    synid = synonym["identifier"] if synonym["identifier"] else ""
+                    if synid == searchterm:
+                        matchlist[synonym["synonymtype"]] = synonym["weight"]
```

Comment:
This seems good but brittle.  Any added type to the document will need to be added to the array or have an `if` block specifically matching.

Perhaps there is a solution where we can instead 

- iterate through all keys
- check the related item's type
  - if string, then `if self.identifiers[id_name] == searchterm:  matchlist[id_name] = 1`
  - if dict, then `if self.identifiers[id_name]['identifiers'] == searchterm:  matchlist[id_name] = self.identifiers[id_name]['weight']`

This would require we codify 'weight' and 'identifier' for "to-many" relationship types (like synonyms).  _Weight could be optional and instead return 1 if there is no value._

**OR**

We could also leave your `if` for synonyms, and just build out the original for loop functionality to be the default for any items that direct match.

```python
for key, value in self.identifiers.items():
    if value == searchterm:
        matchlist[key] = 1
```

Reply:
This deserves a ticket.

Reply:
Sounds like @jtumkur is building it out.